### PR TITLE
[WIP] Asynchronous data loading for long, self-looping simulations

### PR DESCRIPTION
**Pull request by jhrehanoaa**
Still needs to be merged with new updates. Needed to get this in in the condition I'm working with it. Each merge attempt continues to have conflicts. Below is the current output. The main async function has not been properly set to await to second async load function. 

```
creating supernetwork connections set
supernetwork connections set complete
... in 0.2826817035675049 seconds.
organizing connections into reaches ...
reach organization complete
... in 0.1966085433959961 seconds.
setting channel initial states ...
channel initial states complete
... in 6.4373016357421875e-06 seconds.
creating qlateral array ...
qlateral array complete
... in 8.329602241516113 seconds.
executing routing computation ...
Traceback (most recent call last):
  File "compute_nhd_routing_SingleSeg_v02.py", line 1351, in <module>
    loop.run_until_complete(main())
  File "/glade/u/home/jhreha/miniconda3/lib/python3.8/asyncio/base_events.py", line 616, in run_until_complete
    return future.result()
  File "compute_nhd_routing_SingleSeg_v02.py", line 1193, in main
    results = compute_nhd_routing_v02(
  File "compute_nhd_routing_SingleSeg_v02.py", line 734, in compute_nhd_routing_v02
    q0_sub = q0.loc[param_df_sub.index]
AttributeError: 'coroutine' object has no attribute 'loc'
sys:1: RuntimeWarning: coroutine 'load_q0s' was never awaited
```

### CODE REVIEW

**Comment by jhrehanoaa** [comment]
This PR is not showing the changes to compute_nhd_routing_v02, but the changes are in the file. Additionally, the main issue now that I can see is the q0 = load_q0s(ts_iterator,restart_parameters,supernetwork_parameters,param_df) on line 1140 is returning a coroutine object that is unable to be used later in the function.

**Comment by jhrehanoaa** [comment]
As soon as I said that I solved it. 

> This PR is not showing the changes to compute_nhd_routing_v02, but the changes are in the file. Additionally, the main issue now that I can see is the q0 = load_q0s(ts_iterator,restart_parameters,supernetwork_parameters,param_df) on line 1140 is returning a coroutine object that is unable to be used later in the function instead of the normal q0 values. @hellkite500 Are you familiar with this? I am not sure if I need to convert this somehow or if its simply computing incorrectly.
> 
> It will usually say
> 
> ```
> <generator object load_q0s at 0x7f9b844d4ac0>
> ```
> 
> or
> 
> ```
> <coroutine object load_q0s at 0x7f9b844d4ac0>
> ```
> 
> depending on if I use async as a decorator or just an async function.

This can be disregarded. As soon as I said this I actually solved it. It was not computing the aysnc function.

**Comment by jhrehanoaa** [comment]
Restart functionality and async code work currently. Is not started after the to_csv_(restart_file) yet to begin loading for the next iteration. Functionality seems to be working so far though.

File: src/python_routing_v02/compute_nhd_routing_SingleSeg_v02.py [hunk 1]
```diff
@@ -1263,8 +1239,9 @@ def main():
         qvd_columns = pd.MultiIndex.from_product(
             [range(nts), ["q", "v", "d"]]
         ).to_flat_index()
-
         if run_parameters.get("return_courant", False):
+            import pdb
```
Comment (groutr) [inline src/python_routing_v02/compute_nhd_routing_SingleSeg_v02.py:1243]: please remove.

**Comment by jhrehanoaa** [comment]
At its current state, the model runs through the looping and uses the async processes. However, it doesn't look like the process compute speeds aren't changing for me so it is probably still running in serial. Currently, I am calling the ``` async q0 = ``` process after the csv  is written then saving that to a global variable for use on the next run. Does anyone have any suggestions on reordering or restructuring this to make it non-serial? I'm not sure why the task isn't running asynchronously from that point or how it is possible to do so.

File: src/python_routing_v02/compute_nhd_routing_SingleSeg_v02.py [hunk 2]
```diff
@@ -28,7 +28,9 @@
 from joblib import delayed, Parallel
 from itertools import chain, islice
 from operator import itemgetter
-
+import xarray as xr
+import asyncio
+import tracemalloc
```
Comment (jhrehanoaa) [inline src/python_routing_v02/compute_nhd_routing_SingleSeg_v02.py:33]: delete

**Comment by jameshalgren** [comment]
A few comments on the process. Additional code coming... stand by.

Asynchronous execution Psuedocode
   ```
        Sync1: Prepare first warmstate from files
        Sync1: Prepare first forcing from files

        For first forcing set
            Sync2a: run model
            Sync2b: begin preparing next forcing
            Sync3a - AFTER Sync2a, prepare next warmstate (last state of model run)
            Sync3b: write any output from Sync2a
            Loop has to wait for Sync2a+b+Sync3a, does not have to wait for Sync3b
                  if next forcing prepared
```

image

**Comment by jameshalgren** [comment]
Closing -- will potentially reopen for #314.

**Pull request closed without merging**